Fixed a bug in munge_sumstats.py and removed deprecation warnings with new version of pandas

jtb324 opened the pull request:
**Fixing an mistake accidentally introduced in the python 2to3 conversion:**
---
_summary:_ Python3 treats byte objects different than str and this was causing a bug in the line 144 list comprehension of 
munge_sumstats.py

_full explanation of problem:_
There was an error being produced on line 128 of munge_sumstats.py in the list comprehension "return [x.rstrip('\n') for x in opened_file.readline().split()]". This bug was seemingly introduced by the 2to3 conversion. ldsc 1.0.0 call the "openfun" function which returned a callable function based on the provided file compression. When the user provided a gzip file it would follow the if branch that uses gzip.open. The list compreshension on line 144 would then use this callable and call readlines on the resulting filehandle. In python 2.7, readlines returned a byte object which is just an alias to the str type (according to this answer on stackoverflow: [stackoverflow answer](https://stackoverflow.com/questions/5901706/the-bytes-type-in-python-2-7-and-pep-358)). In python 3, byte is a class different from str. This change in byte was causing the x.rstrip in the list comprehension to fail. 

_solution:_
I split the get_compression into two functions: get_compression and open_file. get_compression returns the appropriate compression type for the provided file. This function was not originally used in the above bug but was used later in the code. The open_file function actually opened the file and returned the filehandle. This allowed me to modify gzip.open to "gzip.open(_filehandle_, "rt")" which makes sure that the later call to readlines returns a str object.

**Removing pandas deprecation errors:**
---
_summary:_ There were several errors from pandas from deprecated named arguments and old syntax

_full explanation of problem:_
Newer versions of pandas gave a deprecation error when the delim_whitespace argument was passed to pd.read_csv. There was also an incompatible dtypes error on line 444 of munge_stats.py. This error was cause by the code changing the values of certain indices in a pandas series using an old syntax. 

_solution:_
I changed the named argument from delim_whitespace to 'sep="\t"' as pandas suggested in all the read_csv functions. I also changed the syntax at line 444 of munge_stats.py to a new syntax using loc which suppressed the previous deprecation error.

**Adding dockerfile:**
---
_summary:_ I made a very basic dockerfile that builds from the ubuntu 22.04 image.

_full explanation of what was done:_
People have reported that the make_annot.py script in ldsc fails when ldsc is pip installed because the package pybedtools does not automatically install bedtools. This causes the package pybedtools to not be able to find its dependency bedtools and therefore the make_annot.py script fails. 

_solution:_
I created a very basic docker file that builds off of ubuntu 22.04. This file install bedtools and python3.10 and then install all the python dependencies and copies ldsc. It currently runs on my mac and I build it for both the arm64 and amd64 architectures so it should work on linux servers. I have not been able to test this yet but I hope to soon

**Things left:**
---
There is still an warning left in the code indicating that in future versions of pandas (>=3.0) that pyarrow will be required. I did not feel the need to fix this because pyarrow is not currently used in the code.

jtb324 committed:
There was a bug not caught by the 2to3 python conversion tool. In python2.7 a bytes object was just a string so the gzip.open function returned a string when readlines was called. In python3+, gzip.open by default returns a byte object when readlines is called. This byte object was throwing an error in the list conversion on line 128 because x.rsplit did not work on a byte object. I split the get_compression into two functions: 'get_compression' and 'open_file'. Get compression determines what compression was used on the provided filhandle. 'open_file' opens the file using the appropriate library. The open file function returns the opened file handle which allows us to change the gzip.open function to gzip.open(fh, 'rt') which causes readlines to now return a string.

jtb324 committed:
changed the delim_whitespace=True to sep='\s+'. This change was made to supress a deprecation warning in future versions of pandas. This change was done in the parse.py and minge_sumstats.py file. Also had to made a change to line 444 in munge_sumstats.py. The original code used an outdated syntax of pandas that gave a incompatible dtypes warning. The new code gets rid of this warning and produces a pandas series with the same value counts.

jtb324 committed:
created a very basic docker file built on ubuntu 22.04. This was done because there can be a problem with bedtools in the pip install where the make_annot.py script cannot find bedtools. This image should install bedtools and make everything work. It has not yet been tested. It also needs to be optimized because it is quite big

jtb324 committed:
bumped the patch number on the version to represent a bug fix

jtb324 committed:
Had to update pybedtools to 0.9.1 because there was a AttributeNotFound bug in pybedtools 0.9.0 but the maintainer fixed it in the patch

jtb324 commented:
# 4/10/24 Update:
---
I was able to test the make_annot.py functionality within the docker/singularity image. This script would not work at first because there was a bug in pybedtools 0.9.0. I updated this package to 0.9.1 and the bug is resolved. I have tested the output and it seems to match prior output from ldsc 1.0.0 indicating that the docker image does work.

jtb324 committed:
updated the readme to also explain how to install ldsc using docker and pip. Also explained how there is a bed tools

jtb324 committed:
updated the CHANGELOG to reflect major changes for version 2.0.2

pettyalex commented on dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN  apt-get update \
+  && apt-get install -y wget python3.10 bedtools python3-pip \
+  && rm -rf /var/lib/apt/lists/* \
```
What does this do? New to me.

Also, does pybedtools need wget?

pettyalex commented on dockerfile:
```diff
@@ -0,0 +1,26 @@
+  && rm -rf /var/lib/apt/lists/* \
+  && pip install --upgrade pip
+
+# creating a folder for ldsc
+RUN mkdir ldsc
+
+COPY make_annot.py /app/ldsc/
+COPY munge_sumstats.py /app/ldsc/
+COPY ldsc.py /app/ldsc/
+COPY ./ldscore/ /app/ldsc/ldscore/
+
+# install all of the python dependencies
+COPY requirements.txt /app/
+
+RUN pip install -r requirements.txt
```
In theory we could just use pip to install the package into a venv and activate it instead, right? So that we'd be using the same code for installation both in Docker & PyPI?

pettyalex commented on ldsc.py:
```diff
@@ -27,7 +27,7 @@
     raise ImportError('LDSC requires pandas version >= 0.17.0')
 
-__version__ = '2.0.0'
+__version__ = '2.0.2'
```
Good catch.

pettyalex commented on ldsc.py:
```diff
@@ -599,6 +599,7 @@ def ldscore(args, log):
         log.log(header)
         log.log('Beginning analysis at {T}'.format(T=time.ctime()))
         start_time = time.time()
+        print(args.ref_ld)
```
We should prefer `log`, right?

This is maybe the only change I think we really need to make.

pettyalex commented on munge_sumstats.py:
```diff
@@ -431,7 +441,7 @@ def allele_merge(dat, alleles, log):
     match = a1234.apply(lambda y: y in sumstats.MATCH_ALLELES)
     jj = pd.Series(np.zeros(len(dat), dtype=bool))
-    jj[ii] = match
+    jj.loc[ii] = match # updated this to a modern pandas syntax to get rid of an error about incompatible dtypes
```
I dont think we need the comment here, this is a good fix.

pettyalex reviewed (approved):
Looks good to me.

Was the `print` you left in intentional? Should it be a `log`, or do you want to keep it as is?

jtb324 committed:
removed the unnecessary print statement and then updated the docker image to be slimer by switching to debian-slim and by using system python instead of 3.10

jtb324 commented:
# 4/11/24 - Update
---
**Changes:**
* I removed the print statement
* Updated the docker file to use debian:bulleye-slim and to use system python instead of python 3.10
* Updated setup.py to have the correct pybedtools version
* Also updated image to use pip to localling install ldsc instead of pip installing from requirements.txt

pettyalex commented:
Merged, I'll deploy it tomorrow
